agent/consul/gateways/controller_gateways.go

```diff
@@ -2,201 +2,563 @@ package gateways
-			Topic:   state.EventTopicAPIGateway,
-			Subject: stream.SubjectWildcard,
-		},
-	)
 }
 
-// Reconcile takes in a controller request and ensures this api gateways corresponding BoundAPIGateway exists and is
-// up to date
-func (r *apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
+func reconcileEntry[T structs.ControlledConfigEntry](store *state.Store, ctx context.Context, req controller.Request, reconciler func(ctx context.Context, req controller.Request, store *state.Store, entry T) error, cleaner func(ctx context.Context, req controller.Request, store *state.Store) error) error {
+	_, entry, err := store.ConfigEntry(nil, req.Kind, req.Name, req.Meta)
+	if err != nil {
+		return err
+	}
+	if entry == nil {
+		return cleaner(ctx, req, store)
+	}
+	return reconciler(ctx, req, store, entry.(T))
+}
 
-	r.logger.Debug("started reconciling gateway", "gateway", req.Name)
+func (r apiGatewayReconciler) enqueueCertificateReferencedGateways(store *state.Store, _ context.Context, req controller.Request) error {
+	logger := r.logger.With("inline-certificate", req.Name, "partition", req.Meta.PartitionOrDefault(), "namespace", req.Meta.NamespaceOrDefault())
+	logger.Debug("certificate changed, enqueueing dependent gateways")
+	defer logger.Debug("finished enqueuing gateways")
 
-	metaGateway, err := r.initGatewayMeta(req)
+	_, entries, err := store.ConfigEntriesByKind(nil, structs.APIGateway, acl.WildcardEnterpriseMeta())
 	if err != nil {
 		return err
-	} else if metaGateway == nil {
-		//delete meta gateway
-		r.logger.Info("cleaning up deleted gateway object", "request", req)
-		if err := r.store.Delete(&structs.BoundAPIGatewayConfigEntry{
-			Kind:           structs.BoundAPIGateway,
-			Name:           req.Name,
-			EnterpriseMeta: *req.Meta,
-		}); err != nil {
-			msg := "error cleaning up deleted gateway object"
-			r.logger.Error(msg, err)
-			return errors.Wrap(err, msg)
+	}
+	requests := []controller.Request{}
+	for _, entry := range entries {
+		gateway := entry.(*structs.APIGatewayConfigEntry)
+		for _, listener := range gateway.Listeners {
+			for _, certificate := range listener.TLS.Certificates {
+				if certificate.IsSame(&structs.ResourceReference{
+					Kind:           req.Kind,
+					Name:           req.Name,
+					EnterpriseMeta: *req.Meta,
+				}) {
+					requests = append(requests, controller.Request{
+						Kind: structs.APIGateway,
+						Name: gateway.Name,
+						Meta: &gateway.EnterpriseMeta,
+					})
+				}
+			}
 		}
-		return nil
 	}
+	r.controller.Enqueue(requests...)
+	return nil
+}
 
-	r.ensureBoundGateway(metaGateway)
+func (r apiGatewayReconciler) cleanupBoundGateway(_ context.Context, req controller.Request, store *state.Store) error {
+	logger := r.logger.With("bound-gateway", req.Name, "partition", req.Meta.PartitionOrDefault(), "namespace", req.Meta.NamespaceOrDefault())
+	logger.Debug("cleaning up bound gateway")
+	defer logger.Debug("finished cleaning up bound gateway")
```

Comment:
do we always want to log this? could this be misleading in the case where we bail early and return an error?

Reply:
So, I'm thinking this is good to have for timing more than anything, mainly since the loggers inject timestamps at each case. I think what I'd probably prefer to do is have this, but just ensure that each error code path actually logs the error as well.

Reply:
Going to resolve this @jm96441n since I cleaned up the logging and added error messages wherever we return an error from the reconciliation loop. Let me know if there are any other logging changes you want met to make.